Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,418 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-gsoc open source project
+//
+// Copyright (c) 2022 Apple Inc. and the swift-kafka-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crdkafka
+import struct Foundation.UUID // TODO: can we avoid this Foundation import??
+import Logging
+import NIOCore
+
+/// `NIOAsyncSequenceProducerDelegate` implementation handling backpressure for ``KafkaConsumer``.
+actor ConsumerMessagesAsyncSequenceDelegate: NIOAsyncSequenceProducerDelegate {
+    weak var consumer: KafkaConsumer?
+
+    // TODO: is it ok to block here -> we can make poll non-async and get rid of the task here
+    nonisolated func produceMore() {
+        Task {
+            await self.consumer?.produceMore()
+        }
+    }
+
+    nonisolated func didTerminate() {
+        return
+    }
+
+    func setConsumer(_ consumer: KafkaConsumer) {
+        self.consumer = consumer
+    }
+}
+
+/// `AsyncSequence` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
+public struct ConsumerMessagesAsyncSequence: AsyncSequence {
+    public typealias Element = Result<KafkaConsumerMessage, Error> // TODO: replace with something like KafkaConsumerError
+    typealias HighLowWatermark = NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark
+    let wrappedSequence: NIOAsyncSequenceProducer<Element, HighLowWatermark, ConsumerMessagesAsyncSequenceDelegate>
+
+    /// `AsynceIteratorProtocol` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
+    public struct ConsumerMessagesAsyncIterator: AsyncIteratorProtocol {
+        let wrappedIterator: NIOAsyncSequenceProducer<
+            Element,
+            HighLowWatermark,
+            ConsumerMessagesAsyncSequenceDelegate
+        >.AsyncIterator
+
+        public mutating func next() async -> Element? {
+            await self.wrappedIterator.next()
+        }
+    }
+
+    public func makeAsyncIterator() -> ConsumerMessagesAsyncIterator {
+        return ConsumerMessagesAsyncIterator(wrappedIterator: self.wrappedSequence.makeAsyncIterator())
+    }
+}
+
+/// Receive messages from the Kafka cluster.
+public actor KafkaConsumer {
```

Comment:
Again, chose an `actor` here to have out-of-the-box thread safety for our `NIOAsyncSequenceProducer`. Please double-check how I implemented the connection between the `NIOAsyncSequenceProducer` and the `KafkaConsumer`, maybe there is a better approach.